domain/machine/state/state.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	coredb "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	blockdevice "github.com/juju/juju/domain/blockdevice/state"
+	"github.com/juju/juju/domain/life"
+)
+
+// Logger is the interface used by the state to log messages.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// State describes retrieval and persistence methods for storage.
+type State struct {
+	*domain.StateBase
+	logger Logger
+}
+
+// NewState returns a new state reference.
+func NewState(factory coredb.TxnRunnerFactory, logger Logger) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// UpsertMachine creates or updates the specified machine.
+// TODO - this just creates a minimal row for now.
+func (s *State) UpsertMachine(ctx context.Context, machineId string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return upsertMachine(ctx, tx, machineId)
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.logger.Debugf("upserted machine %q", machineId)
+
+	return nil
+}
+
+func upsertMachine(ctx context.Context, tx *sqlair.TX, machineId string) error {
```

Comment:
I think extracting this for re-use as a bootstrap concern is taking DRY too far.

To achieve it, we're pushing a slew of logic into a write transaction (statement preparations for example), but at bootstrap time we know there won't be any conflicts, as the machine can't exist - it's never an upsert.

I'd rather duplicate the minimum required for a bootstrap machine creation, and keep the more frequent case lean.

Reply:
Refactored it